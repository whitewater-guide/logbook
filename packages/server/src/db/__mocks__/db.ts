import {
  DatabasePoolConnectionType,
  DatabasePoolType,
  createPool,
} from 'slonik';

import { Client } from 'pg';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: `logbook_test_${process.env.JEST_WORKER_ID}`,
  password: process.env.POSTGRES_PASSWORD,
};

let _pool: DatabasePoolType | undefined;

export const db: () => DatabasePoolConnectionType = () => {
  if (!_pool) {
    throw new Error('no test pool found');
  }
  return _pool;
};

export const setupDB = async () => {
  const workerID = process.env.JEST_WORKER_ID;
  if (!workerID) {
    throw new Error('JEST_WORKER_ID not found');
  }
  const client = new Client({
    host: process.env.POSTGRES_HOST,
    user: process.env.POSTGRES_USER,
    database: process.env.POSTGRES_DB,
    password: process.env.POSTGRES_PASSWORD,
  });
  try {
    await client.connect();
    await client.query(`DROP DATABASE IF EXISTS logbook_test_${workerID}`);
    await client.query(
      `CREATE DATABASE logbook_test_${workerID} TEMPLATE logbook_test_template`,
    );
  } finally {
    await client.end();
  }

  _pool = createPool(
    `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  );
};

export const teardownDB = async () => {
  await _pool?.end();
};
