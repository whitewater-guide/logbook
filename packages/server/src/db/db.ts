import { createPool, DatabasePoolConnectionType } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
