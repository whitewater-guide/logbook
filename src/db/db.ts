import {
  DatabasePoolConnectionType,
  createPool,
  createTypeParserPreset,
} from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
