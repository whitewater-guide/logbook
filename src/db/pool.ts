import { createPool, createTypeParserPreset } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database:
    process.env.NODE_ENV === 'test'
      ? `logbook_test_${process.env.JEST_WORKER_ID}`
      : process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

export const db = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);
