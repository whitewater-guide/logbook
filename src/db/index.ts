import { db } from './pool';

export default db;
