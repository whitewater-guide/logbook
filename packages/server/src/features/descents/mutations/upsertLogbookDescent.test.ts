import {
  DESCENT_1,
  DESCENT_2,
  DESCENT_2_SHARE_TOKEN,
  DESCENT_4_SHARE_TOKEN,
  SECTION_1,
  SECTION_2,
  USER_1,
  USER_2,
} from '~/test/fixtures';
import {
  UpsertLogbookDescentMutation,
  UpsertLogbookDescentMutationVariables,
} from './upsertLogbookDescent.test.generated';
import { db, setupDB, teardownDB } from '~/db';

import { LogbookDescentInput } from '~/__generated__/graphql';
import { gql } from 'apollo-server';
import { runQuery } from '~/test/apollo-helpers';
import { sql } from 'slonik';

beforeEach(setupDB);
afterEach(teardownDB);

const mutation = gql`
  mutation upsertLogbookDescent(
    $descent: LogbookDescentInput!
    $shareToken: String
  ) {
    upsertLogbookDescent(descent: $descent, shareToken: $shareToken) {
      id
      userId

      startedAt
      duration
      level {
        value
        unit
      }
      comment
      public

      upstreamData

      createdAt
      updatedAt

      section {
        id

        region
        river
        section
        difficulty
        putIn {
          lat
          lng
        }
        takeOut {
          lat
          lng
        }

        upstreamId
        upstreamData

        createdAt
        updatedAt
      }
    }
  }
`;

const descent: LogbookDescentInput = {
  section: {
    region: 'Georgia',
    river: 'Bzhuzha',
    section: 'Lower',
    difficulty: 2,
  },
  startedAt: new Date(2000, 1, 1),
  duration: 3600,
  comment: 'comment',
  level: {
    value: 20,
    unit: 'cfs',
  },
  public: true,
  upstreamData: { foo: 'bar' },
};

it('anon should fail to insert descent', async () => {
  const result = await runQuery<
    UpsertLogbookDescentMutation,
    UpsertLogbookDescentMutationVariables
  >(mutation, { descent });
  expect(result.errors).toBeDefined();
  expect(result.data.upsertLogbookDescent).toBeNull();
});

it('non-owner should fail to update descent', async () => {
  const result = await runQuery<
    UpsertLogbookDescentMutation,
    UpsertLogbookDescentMutationVariables
  >(mutation, { descent: { ...descent, id: DESCENT_1 } }, USER_2);
  expect(result.errors).toBeDefined();
  expect(result.data.upsertLogbookDescent).toBeNull();
});

it('should fail on validation check', async () => {
  const badLogbookDescent: LogbookDescentInput = {
    section: {
      region: 'G',
      river: 'B',
      section: 'L',
      difficulty: 222,
    },
    startedAt: new Date(2000, 1, 1),
  };
  const result = await runQuery<
    UpsertLogbookDescentMutation,
    UpsertLogbookDescentMutationVariables
  >(mutation, { descent: badLogbookDescent }, USER_1);
  expect(result.errors?.[0]?.extensions).toMatchSnapshot({
    id: expect.any(String),
  });
  expect(result.data.upsertLogbookDescent).toBeNull();
});

it('should insert', async () => {
  const result = await runQuery<
    UpsertLogbookDescentMutation,
    UpsertLogbookDescentMutationVariables
  >(mutation, { descent }, USER_1);
  expect(result.errors).toBeUndefined();
  const { section, ...rest } = result.data.upsertLogbookDescent!;
  expect(section).toMatchSnapshot<any>({
    createdAt: expect.any(Date),
    updatedAt: expect.any(Date),
    id: expect.any(String),
  });
  expect(rest).toMatchSnapshot<any>({
    createdAt: expect.any(Date),
    updatedAt: expect.any(Date),
    id: expect.any(String),
  });
});

it('should update', async () => {
  const result = await runQuery<
    UpsertLogbookDescentMutation,
    UpsertLogbookDescentMutationVariables
  >(
    mutation,
    {
      descent: {
        ...descent,
        id: DESCENT_1,
        section: { ...descent.section, id: SECTION_1 },
      },
    },
    USER_1,
  );
  expect(result.errors).toBeUndefined();
  const { section, ...rest } = result.data.upsertLogbookDescent!;
  expect(section).toMatchSnapshot<any>({
    updatedAt: expect.any(Date),
  });
  expect(rest).toMatchSnapshot<any>({
    updatedAt: expect.any(Date),
  });
});

type ParentTestCase = [string, string, string, string];

it.each<ParentTestCase>([
  [' ', DESCENT_2_SHARE_TOKEN, DESCENT_2, SECTION_2],
  [' recursive ', DESCENT_4_SHARE_TOKEN, DESCENT_1, SECTION_1],
])(
  'should set%sparent references when token is provided',
  async (
    _,
    shareToken,
    expectedParentLogbookDescentId,
    expectedParentLogbookSectionId,
  ) => {
    const result = await runQuery<
      UpsertLogbookDescentMutation,
      UpsertLogbookDescentMutationVariables
    >(
      mutation,
      {
        descent,
        shareToken,
      },
      USER_2,
    );
    expect(result.errors).toBeUndefined();
    const id = result.data.upsertLogbookDescent?.id;
    const sectionId = result.data.upsertLogbookDescent?.section.id;
    const parentDescId = await db().oneFirst(
      sql`SELECT parent_id FROM logbook_descents WHERE id = ${id}`,
    );
    const parentLogbookSectionId = await db().oneFirst(
      sql`SELECT parent_id FROM logbook_sections WHERE id = ${sectionId}`,
    );
    expect(parentDescId).toBe(expectedParentLogbookDescentId);
    expect(parentLogbookSectionId).toBe(expectedParentLogbookSectionId);
  },
);
