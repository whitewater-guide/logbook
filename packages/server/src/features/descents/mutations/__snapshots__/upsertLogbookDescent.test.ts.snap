// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail on validation check 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "stacktrace": undefined,
    "validationErrors": Object {
      "descent": Object {
        "section": Object {
          "difficulty": Object {
            "key": "yup:number.max",
            "options": Object {
              "max": 6,
              "path": "descent.section.difficulty",
            },
          },
          "region": Object {
            "key": "yup:string.min",
            "options": Object {
              "min": 2,
              "path": "descent.section.region",
            },
          },
          "river": Object {
            "key": "yup:string.min",
            "options": Object {
              "min": 2,
              "path": "descent.section.river",
            },
          },
          "section": Object {
            "key": "yup:string.min",
            "options": Object {
              "min": 2,
              "path": "descent.section.section",
            },
          },
        },
      },
    },
  },
  "id": Any<String>,
  "validationErrors": Object {
    "descent": Object {
      "section": Object {
        "difficulty": Object {
          "key": "yup:number.max",
          "options": Object {
            "max": 6,
            "path": "descent.section.difficulty",
          },
        },
        "region": Object {
          "key": "yup:string.min",
          "options": Object {
            "min": 2,
            "path": "descent.section.region",
          },
        },
        "river": Object {
          "key": "yup:string.min",
          "options": Object {
            "min": 2,
            "path": "descent.section.river",
          },
        },
        "section": Object {
          "key": "yup:string.min",
          "options": Object {
            "min": 2,
            "path": "descent.section.section",
          },
        },
      },
    },
  },
}
`;

exports[`should insert 1`] = `
Object {
  "createdAt": Any<Date>,
  "difficulty": 2,
  "id": Any<String>,
  "putIn": null,
  "region": "Georgia",
  "river": "Bzhuzha",
  "section": "Lower",
  "takeOut": null,
  "updatedAt": Any<Date>,
  "upstreamData": null,
  "upstreamId": null,
}
`;

exports[`should insert 2`] = `
Object {
  "comment": "comment",
  "createdAt": Any<Date>,
  "duration": 3600,
  "id": Any<String>,
  "level": Object {
    "unit": "cfs",
    "value": 20,
  },
  "public": true,
  "startedAt": 2000-02-01T00:00:00.000Z,
  "updatedAt": Any<Date>,
  "upstreamData": Object {
    "foo": "bar",
  },
  "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
}
`;

exports[`should update 1`] = `
Object {
  "createdAt": 2020-01-01T00:00:00.000Z,
  "difficulty": 2,
  "id": "45e70501-5d20-449f-b854-52b06b138831",
  "putIn": null,
  "region": "Georgia",
  "river": "Bzhuzha",
  "section": "Lower",
  "takeOut": null,
  "updatedAt": Any<Date>,
  "upstreamData": null,
  "upstreamId": null,
}
`;

exports[`should update 2`] = `
Object {
  "comment": "comment",
  "createdAt": 2020-01-01T00:00:00.000Z,
  "duration": 3600,
  "id": "97998f67-74f2-43db-a4e4-198c2ca540a7",
  "level": Object {
    "unit": "cfs",
    "value": 20,
  },
  "public": true,
  "startedAt": 2000-02-01T00:00:00.000Z,
  "updatedAt": Any<Date>,
  "upstreamData": Object {
    "foo": "bar",
  },
  "userId": "ad808d80-7d1d-4bee-8276-0b9204e46b2f",
}
`;
